### PR TITLE
fix(vscode): workspace-trust gating + exec timeouts + exclusions + activation (#348 VS Code)

### PR DESCRIPTION
**Pull request by jainal09**
## Summary

Five VS Code extension fixes (the `#348` VS Code findings, including the folded `#363`/`#364`), focused on **workspace-trust safety**.

### V1 — workspace trust (runtime + manifest)
The extension auto-encrypts on file close (enabled by default) and ran `envdrift lock` against **untrusted** workspace content with no trust check. Now:
- the close-watcher is registered only when `vscode.workspace.isTrusted` (and re-registered on `onDidGrantWorkspaceTrust`), with a hard `isTrusted` guard in `handleDocumentClose` **and** the manual command;
- `package.json` declares `capabilities.untrustedWorkspaces: "limited"` with `envdrift.enabled` as a `restrictedConfiguration` — so status/agent UI still work untrusted, but the binary never runs against untrusted content.

### V2 (#363) — exec timeouts
All agent `execAsync` calls — `--version` (×2) and the recurring **30s** status poll — now pass `{ timeout: 10000 }`, matching start/stop. A hung `envdrift-agent` can no longer accumulate stuck polling processes.

### V3 — manual-encrypt exclusions
The manual encrypt command checked patterns but not exclusions; you could hand-encrypt `.env.keys` (the decryption key → lockout) or `.env.example`. Now it applies the same `isExcluded` gate as the auto path.

### V4 — real pattern matching
`isExcluded` did exact-basename `Array.includes` despite being advertised as patterns. Now it globs (shared with `matchesPatterns`), so `*.local` / `.env.*` / path excludes work.

### V5 — activation
`activationEvents` narrowed from unconditional `onStartupFinished` (watchers + 30s polling in **every** workspace) to `workspaceContains:**/.env*` / `**/envdrift.toml` + the extension's commands.

## Verification
- `npx tsc --noEmit` clean; `eslint` clean.
- **43 unit tests pass** (`mocha`) — glob matching (literal dot, `*.local`, `.env.*`, negatives) and the V3 manual-exclude gate.
- Manifest validated (`untrustedWorkspaces: limited`, narrowed `activationEvents`).
- Note: full UI behavior (trust gating, activation triggers) needs a VS Code host (`@vscode/test-electron`) and was verified by code review — the pure logic (glob/exclude) is unit-tested directly.

Addresses the VS Code findings of #348 (incl. folded #363/#364).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR delivers five targeted VS Code extension fixes addressing workspace-trust safety, exec timeouts, exclusion enforcement on the manual-encrypt path, real glob semantics for exclusions, and a narrowed activation footprint.

- **Trust gating (V1):** The close-watcher is now registered only in trusted workspaces (with lazy re-registration on `onDidGrantWorkspaceTrust`), and both `handleDocumentClose` and `encryptCurrentFile` carry hard `isTrusted` guards; `package.json` adds `capabilities.untrustedWorkspaces: "limited"`.
- **Exec timeouts (V2):** A shared `AGENT_EXEC_TIMEOUT_MS = 10 000` constant is applied uniformly to all five `execAsync` call sites in `agentStatus.ts`, preventing hung polling processes.
- **Exclusion & glob fixes (V3/V4):** The manual-encrypt command now calls `isExcluded` before proceeding, and `isExcluded` itself is upgraded from exact `Array.includes` to the same `matchesGlob` helper used by `matchesPatterns`, so wildcard patterns like `*.keys` and `.env.*` work correctly.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all five fixes are narrowly scoped, well-guarded, and covered by 43 unit tests.

The trust-gating logic is applied at both the registration site and each handler entry point, so there is no path that reaches encryptFile or the envdrift binary in an untrusted workspace. The exec-timeout change is a straight constant extraction with no behavioral regression. The isExcluded glob upgrade correctly escapes all regex specials and the test suite explicitly validates dot-as-literal, *.local, .env.*, and combined exclude lists. No pre-existing correct behavior is removed or weakened.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| envdrift-vscode/src/fileWatcher.ts | Adds workspace trust gating to both auto-encrypt (handleDocumentClose) and manual-encrypt (encryptCurrentFile) paths, and adds the missing isExcluded check to the manual path. Logic is correct: guard is at entry, and re-registration on trust grant is idempotent. |
| envdrift-vscode/src/utils.ts | Extracts matchesGlob helper and upgrades isExcluded from exact Array.includes to glob-pattern matching. Regex escaping correctly handles all special chars while leaving * for wildcard expansion. Basename extraction is consistent in both functions. |
| envdrift-vscode/src/agentStatus.ts | Introduces AGENT_EXEC_TIMEOUT_MS constant (10 000 ms) and applies it uniformly to all five execAsync call sites, including the previously-uncovered --version (x2) and status poll calls. Deduplication of the magic literal 10000 is clean. |
| envdrift-vscode/package.json | Narrows activationEvents from the unconditional onStartupFinished to workspaceContains and onCommand entries for all 7 contributed commands; adds capabilities.untrustedWorkspaces with supported:'limited' and envdrift.enabled as a restrictedConfiguration. |
| envdrift-vscode/src/test/unit/exclusion.test.ts | New test file comprehensively covers isExcluded glob semantics: exact names, *.local suffix glob, .env.* prefix glob, path inputs (basename extraction), dot-literal correctness, and mixed exclude lists with negatives. |
| envdrift-vscode/src/test/unit/shouldEncrypt.test.ts | New test file mirrors the manual-encrypt gate (matchesPatterns && !isExcluded) added in V3, verifying that .env.example, *.local, and .env.keys are blocked even when explicitly invoked. |
| envdrift-vscode/src/test/unit/config.test.ts | Extends existing config tests with wildcard, prefix+suffix, and literal-as-exact-match cases for isExcluded, covering the regex-escaping edge (dot is literal, not any-char). |

</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `envdrift-vscode/src/agentStatus.ts`, line 37-44 ([link](https://github.com/jainal09/envdrift/blob/a35f9b0f6d0429ae092a461c62183627f0b26893/envdrift-vscode/src/agentStatus.ts#L37-L44)) 

   <a href="#">P2</a> `checkAgentStatus` calls `isAgentInstalled()` (which spawns `envdrift-agent --version`) and then calls `getAgentVersion()` (a second `--version` spawn) for every `running` or `stopped` result. Each 30-second poll therefore fires `--version` twice. This is a pre-existing pattern, but the timeout fix makes it a good moment to consolidate — capturing the version once from `isAgentInstalled` avoids the double execution.

   

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!
</details>

<sub>Reviews (3): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/vscode-ext"](https://github.com/jainal09/envdrift/commit/9eae18f1a7b88581e7e375f1b679e66fe5523623) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=35988967)</sub>

### CODE REVIEW

File: envdrift-vscode/package.json [hunk 1]
```diff
@@ -24,9 +24,24 @@
...
   "activationEvents": [
-    "onStartupFinished"
+    "workspaceContains:**/.env",
+    "workspaceContains:**/.env.*",
+    "workspaceContains:**/envdrift.toml",
+    "onCommand:envdrift.enable",
+    "onCommand:envdrift.encryptNow",
+    "onCommand:envdrift.startAgent",
+    "onCommand:envdrift.refreshAgentStatus"
   ],
```
[the remark this answers is not in the document]
Comment (jainal09) [reply envdrift-vscode/package.json:37]: Fixed in 2396c01: added `onCommand:` activation events for all 7 contributed commands (the V5 narrowing had missed disable/showStatus/stopAgent).

**Comment by jainal09** [comment]
🤖 **envdrift review-bot** — automated re-review of commit 2396c01


The VS Code extension PR lands a security-critical workspace-trust gate plus exec-timeout and activation-narrowing changes, but ships no regression test for the headline trust gate and updates no docs. Both findings are omissions (no diff RIGHT-side line to anchor to), so they're listed below rather than inline. The previously-flagged activation-events regression (greptile P1) is already fixed in 2396c01.

**Findings (2):**
- [high] Headline workspace-trust fix (V1) and the timeout/activation fixes ship no regression test — `envdrift-vscode/src/test/suite/extension.test.ts` (see below — not anchored)
- [medium] VS Code extension docs not updated for trust-gating / activation behavior changes — `docs/guides/vscode-extension.md` (see below — not anchored)

### Not anchored inline

These findings concern omitted code/docs (no added line exists to attach an inline comment to), so they are reported here in full.

---

#### [high · test-gap] Headline workspace-trust fix (V1) and the timeout/activation fixes ship no regression test
`envdrift-vscode/src/test/suite/extension.test.ts`

This PR's headline change is the workspace-trust security gate (V1, addressing #348 and folding in #364): `registerCloseWatcher` is only registered when `vscode.workspace.isTrusted`, with hard `if (!vscode.workspace.isTrusted) return;` guards in `handleDocumentClose` and a warning-and-return guard in `encryptCurrentFile`. None of these guards are covered by any test.

The new unit tests only exercise V3 (manual-encrypt exclude gate) and V4 (glob matching): `shouldEncrypt.test.ts` re-implements the gate as `matchesPatterns(name) && !isExcluded(name)` in the test file rather than driving `fileWatcher.encryptCurrentFile`, so it cannot catch a missing trust guard; `exclusion.test.ts`/`config.test.ts` only test the pure `isExcluded` glob. The integration suite `src/test/suite/extension.test.ts` is unchanged and only asserts extension presence, activation, command registration, and config defaults.

A VS Code host is available in CI: `vscode-ci.yml` runs `npm test` (`@vscode/test-electron` via `runTest.ts`/`suite/index.ts`) under xvfb on Linux/macOS/Windows. So the trust gate (e.g. assert `encryptCurrentFile` shows the untrusted warning / does not call `encryptFile` when `isTrusted` is false, and that the close-watcher is not registered when untrusted) IS testable there. As written, deleting every `isTrusted` guard would leave the entire suite green — i.e. the regression the PR exists to prevent is not protected.

This violates CLAUDE.md: "Every bug fix ships a REAL regression test that FAILS without the fix and PASSES with it." The same gap applies to V2 (exec timeouts) and V5 (activationEvents narrowing). Suggested direction: add a trust-gating test in the test-electron suite (the strongest), at minimum for the manual-encrypt path; do not just re-derive the gate in a standalone helper.

_Evidence:_ extension.test.ts only has: 'Extension should be present', 'Extension should activate', 'Commands should be registered', 'Configuration should have default values' — no isTrusted assertions. fileWatcher.ts adds `if (!vscode.workspace.isTrusted) { return; }` (handleDocumentClose) and `if (!vscode.workspace.isTrusted) { vscode.window.showWarningMessage(...); return; }` (encryptCurrentFile), neither driven by any test. shouldEncrypt.test.ts defines its own `const shouldEncrypt = (name, patterns, exclude) => matchesPatterns(name, patterns) && !isExcluded(name, exclude)` instead of calling the real handler.

---

#### [medium · doc-mismatch] VS Code extension docs not updated for trust-gating / activation behavior changes
`docs/guides/vscode-extension.md`

The PR changes three user-facing behaviors but touches no docs, violating CLAUDE.md's "Keep docs in sync — update docs/ in the same PR as the behavior change."

1. Workspace trust (V1): encryption is now disabled (auto and manual) in untrusted workspaces, and the manual command surfaces `EnvDrift: cannot encrypt — this workspace is not trusted.`. `docs/guides/vscode-extension.md` never mentions workspace trust — not in 'Security Considerations', not in 'Troubleshooting > Files not being encrypted', and not in 'Extension not activating'. A user whose encrypt silently no-ops (auto) or errors (manual) because the folder is untrusted has zero doc guidance.

2. Activation narrowing (V5): activation moved from unconditional `onStartupFinished` to `workspaceContains:**/.env*` / `**/envdrift.toml` + per-command triggers. The 'Troubleshooting > Extension not activating' section gives no hint that a workspace with no matching file (and no command invoked) will not activate.

3. Stale timeout claim: the 'Performance' section states "30-second timeout prevents hung operations", but the agent exec timeout introduced/standardized here is 10s (`AGENT_EXEC_TIMEOUT_MS = 10000`); 30s is the poll interval (`CHECK_INTERVAL_MS`), not a timeout. This pre-existing inaccuracy is now more clearly wrong given the explicit timeout work in V2.

Suggested direction: add a short 'Workspace Trust' note (auto + manual encryption require a trusted workspace; status/agent UI still work), document the new activation triggers in the troubleshooting section, and correct the 10s-timeout vs 30s-interval wording.

_Evidence:_ docs/guides/vscode-extension.md 'Security Considerations' lists only: encrypts on close, uses envdrift lock, keys in .env.keys/vault, ephemeral keys — no trust mention. 'Performance' says '30-second timeout prevents hung operations' while agentStatus.ts sets `const AGENT_EXEC_TIMEOUT_MS = 10000;` (timeout) and `const CHECK_INTERVAL_MS = 30000;` (interval). package.json activationEvents changed from `onStartupFinished` to `workspaceContains:**/.env` + `**/.env.*` + `**/envdrift.toml` + onCommand entries. PR diff contains no docs/ changes.

**Comment by jainal09** [comment]
@coderabbitai review

**Comment by jainal09** [comment]
@coderabbitai review
